pkg/provisioner/fss/fss.go

```diff
@@ -84,175 +95,373 @@ func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, moun
+	fs, err := fsp.getFileSystemByDisplayName(ctx, ad, displayName)
+	if err != nil && err != errNotFound {
+		return nil, err
+	}
+	if fs != nil {
+		return fsp.awaitFileSystem(ctx, logger, *fs.Id)
+	}
+
+	resp, err := fsp.client.FSS().CreateFileSystem(ctx, filestorage.CreateFileSystemRequest{
+		CreateFileSystemDetails: filestorage.CreateFileSystemDetails{
 			AvailabilityDomain: &ad,
-			SubnetId:           &subnetID,
 			CompartmentId:      common.String(fsp.client.CompartmentOCID()),
-			DisplayName:        mtDisplayName,
+			DisplayName:        &displayName,
 		},
 	})
 	if err != nil {
 		return nil, err
 	}
-	return &resp.MountTarget, nil
+
+	logger.With("fileSystemID", *resp.FileSystem.Id).Info("Created filesystem")
+
+	return fsp.awaitFileSystem(ctx, logger, *resp.FileSystem.Id)
+}
+
+// findExport looks for an existing export with the same filesystem ID, export set ID, and path.
+// NOTE: No two non-'DELETED' export resources in the same export set can reference the same file system.
+func (fsp *filesystemProvisioner) findExport(ctx context.Context, fsID, exportSetID string) (*filestorage.ExportSummary, error) {
+	var page *string
+	for {
+		resp, err := fsp.client.FSS().ListExports(ctx, filestorage.ListExportsRequest{
+			CompartmentId: common.String(fsp.client.CompartmentOCID()),
+			FileSystemId:  &fsID,
+			ExportSetId:   &exportSetID,
+			Page:          page,
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, export := range resp.Items {
+			if export.LifecycleState == filestorage.ExportSummaryLifecycleStateCreating ||
+				export.LifecycleState == filestorage.ExportSummaryLifecycleStateActive {
+				return &export, nil
+			}
+		}
+		if page = resp.OpcNextPage; resp.OpcNextPage == nil {
+			break
+		}
+	}
+
+	return nil, errNotFound
+}
+
+func (fsp *filesystemProvisioner) getOrCreateExport(ctx context.Context, logger *zap.SugaredLogger, fsID, exportSetID string) (*filestorage.Export, error) {
+	export, err := fsp.findExport(ctx, fsID, exportSetID)
+	if err != nil && err != errNotFound {
+		return nil, err
+	}
+	if export != nil {
+		return fsp.awaitExport(ctx, logger, *export.Id)
+	}
+
+	path := "/" + fsID
+
+	// If export doesn't already exist create it.
+	resp, err := fsp.client.FSS().CreateExport(ctx, filestorage.CreateExportRequest{
+		CreateExportDetails: filestorage.CreateExportDetails{
+			ExportSetId:  &exportSetID,
+			FileSystemId: &fsID,
+			Path:         &path,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	logger.With("exportID", *resp.Export.Id).Info("Created Export")
+	return fsp.awaitExport(ctx, logger, *resp.Export.Id)
 }
 
 func (fsp *filesystemProvisioner) Provision(options controller.VolumeOptions, ad *identity.AvailabilityDomain) (*v1.PersistentVolume, error) {
 	ctx := context.Background()
-	fsDisplayName := common.String(fmt.Sprintf("%s%s", provisioner.GetPrefix(), options.PVC.Name))
+	fsDisplayName := fmt.Sprintf("%s%s", provisioner.GetPrefix(), options.PVC.UID)
```

Comment:
NOTE: This changes the display name to a UID so that we can uniquely identify existing resources for idempotency.